### ISSUE TITLE
Allow missing docstrings when overriding methods

### ISSUE THREAD

**Issue by torfsen** [PyCQA/pydocstyle#309]
In Python, abstract base classes often play the role of an interface:

```python
class MyInterface:
    def some_method(self):
        "Do something useful"
        pass

class MyImplementation(MyInterface):
    def some_method(self):
        # Actual implementation
```

In such situations, the docstring of the abstract method is often also appropriate for the implementing method, and duplicating it there would be redundant.

However, omitting the docstring causes `pydocstyle` 2.1.1 to emit a D102 warning (missing docstring in public method).

Excluding D102 in general is not a good option, since one obviously wants to catch "normal" undocumented public methods.

I therefore suggest the following feature: if a public method has no docstring then it is first checked whether it overrides an inherited method. If this is not the case then D102 is raised as before. For an overriding method, we check whether one of the parent implementations (there could be multiple in the inheritance chain) has a docstring. If that is the case, a newly created warning is raised (e.g. "DXXX Overriding method inherits docstring from superclass implementation"). If the parent implementations also do not have docstrings then D102 is raised for the overriding method.

This would make sure that, by default, all undocumented public methods are reported (just as they are now). However, inheriting docstrings could be explicitly allowed by excluding the new warning (DXXX).

An obvious work-around in the meantime is to add `# noqa: D102` to the offending methods.

**Comment by sigmavirus24** [comment]
I seem to remember that docstrings aren't inherited though. In other words `MyImplementation.some_method` when implemented doesn't have `"Do something useful"` as its docstring. Is that correct? If so, then you do need a docstring

**Comment by Nurdok** [comment]
@torfsen I would love it if pydocstyle could implement this behavior, unfortunately, it's *very* difficult. Pydocstyle is a static linter, and determining the parent class is pretty complicated without actually running the code.

It would be great if there was some 3rd party implementation of e.g., a decorator which you can use to annotate overridden method like so:

```.py
class MyImplementation(MyInterface):
    @override
    def some_method(self):
        # Actual implementation
```

This could copy the docstring to the overriding method and allow pydocstyle to not emit an error for a missing docstring. The problem is that this is also pretty hard to implement cleanly and there's no wide use of such a decorator.

**Comment by torfsen** [comment]
@sigmavirus24 You are correct, docstrings are not inherited by Python itself:
```
In [1]: class A:
   ...:     def foo(self):
   ...:         """A docstring"""
   ...:         pass
   ...:     

In [2]: class B(A):
   ...:     def foo(self):
   ...:         print("Overridden method")
   ...:         

In [3]: B().foo.__doc__ is None
Out[3]: True
```
Since Python 3.5, [`inspect.getdoc`](https://docs.python.org/3/library/inspect.html#inspect.getdoc) does look up missing docstrings in base classes:
```
In [6]: inspect.getdoc(B().foo)
Out[6]: 'A docstring'
```
The Sphinx autodoc extension [also implements this behavior](http://www.sphinx-doc.org/en/master/ext/autodoc.html#confval-autodoc_inherit_docstrings) (since version 1.7).

In my opinion, while not being a core Python feature, docstring inheritance is still very useful and supported by many relevant tools -- that's why I would love to have support for it in pydocstyle, too.

@Nurdok Thanks for your insights regarding the difficulties of such a feature via static analysis -- to be honest I hadn't thought about how this would be implemented, I simply liked the idea :wink:

You're suggestion of using an `@override` decorator is also very interesting. Upon further research I've found [a package that provides exactly that](https://github.com/mkorpela/overrides). There's also [a detailed discussion on python-ideas](https://mail.python.org/pipermail/python-ideas/2016-July/041095.html) regarding the topic.

In conclusion, I still think that docstring inheritance would be a nice feature to support in pydocstyle, but the [`overrides` package](https://github.com/mkorpela/overrides) is a good alternative.

I'm leaving this issue open in case you want to keep it around as a feature request, but feel free to close it.

**Comment by wulmer** [comment]
> but the overrides package is a good alternative.

I am not sure if I understood this correctly: for what is the `overrides` package a good alternative? For copying docstrings? As you mentioned, this is done by `inspect.getdoc` since Python 3.5.

The main problem I have is that I don't want to add the `# noqa: D102` to every method of my classes. I first thought that adding `@overrides` to the methods makes pydocstyle understand that there is an inheritable docstring, but pydocstyle still complains about the missing docstring.

If the `overrides` decorator added an `__overrides__` attribute to the methods which referred to the correct base class, it would be easy to make `pydocstyle` check the base class for a docstring and then complain if none was found.

This, however, would require a small change in the `overrides` package (adding an `__overrides__` attribute to overriding methods) and a change in `pydocstyle` to make use of the `__overrides__` attribute (to check method and parent class method for docstrings).

The changes shouldn't be too complicated, so my main question is: do you see a need for this feature?

**Comment by sigmavirus24** [comment]
@wulmer those attributes aren't visible to pydocstyle. Those are added at runtime. Pydocstyle parses the files statically which pre-empts `@override` from adding those attributes so we can't inspect them.

**Comment by thomkeh** [comment]
@wulmer in case you didn't know: if you trust the programmers to use the decorator `@overrides` only when the parent class has a docstring, then you can solve the problem by using `@overrides` and specifying `--ignore-decorator=overrides` when you run pydocstyle.

**Comment by randolf-scholz** [comment]
Pylint is able to detect docstrings in the parent class, so another solution is to just disable `D102` in `flake8` and add `--fail-on C0116` option to `pylint`.

**Comment by davisford** [comment]
> @wulmer in case you didn't know: if you trust the programmers to use the decorator `@overrides` only when the parent class has a docstring, then you can solve the problem by using `@overrides` and specifying `--ignore-decorator=overrides` when you run pydocstyle.

@wulmer is the `@override` decorator the one in [PEP 0698](https://peps.python.org/pep-0698/)?  That is in Python 3.12?  I have Python 3.11, and I don't have it.  Sorry if this is a dumb question -- not a regular Python dev.

**Comment by tmke8** [comment]
You can install the `typing-extensions` package and then import it with `from typing_extensions import override`. This will work on any Python version, not just 3.12.

**Comment by davisford** [comment]
@tmke8 thanks!  i got that now.  `pydocstyle` is still complaining about `D102` tho?  This is how I set it in vscode:

![Screen Shot 2023-02-28 at 5 06 02 PM](https://user-images.githubusercontent.com/186677/221992329-ffda9b1a-8fc5-4b31-b1dd-0f1083f8170c.png)


![Screen Shot 2023-02-28 at 5 07 38 PM](https://user-images.githubusercontent.com/186677/221992464-94b5b4d2-7105-4810-94b0-185ff80d62dd.png)
